pthread/pthread.c

```diff
@@ -603,37 +818,95 @@ int pthread_condattr_getclock(const pthread_condattr_t *restrict attr, clockid_t
 
 int pthread_cond_init(pthread_cond_t *restrict cond, const pthread_condattr_t *restrict attr)
 {
-	return condCreate(cond);
+	cond->initialized = 1;
+	return condCreate(&cond->condh);
 }
 
 
 int pthread_cond_destroy(pthread_cond_t *cond)
 {
-	return resourceDestroy(*cond);
+	int err = EOK;
+	mutexLock(pthread_common.mutex_cond_init_lock);
+	if (cond->initialized == 1) {
+		mutexUnlock(pthread_common.mutex_cond_init_lock);
+		err = resourceDestroy(cond->condh);
+	}
+	return err;
 }
 
 
 int pthread_cond_signal(pthread_cond_t *cond)
 {
-	return condSignal(*cond);
+	int err = EOK;
+
+	if (cond->initialized == 0) {
+		mutexLock(pthread_common.mutex_cond_init_lock);
+		if (cond->initialized == 0) {
+			err = condCreate(&cond->condh);
+			if (err == EOK) {
+				cond->initialized = 1;
+			}
+		}
+		mutexUnlock(pthread_common.mutex_cond_init_lock);
+	}
+	if (err == EOK) {
+		err = condSignal(cond->condh);
+	}
+	return err;
 }
 
 
 int pthread_cond_broadcast(pthread_cond_t *cond)
 {
-	return condBroadcast(*cond);
+	int err = EOK;
+
+	if (cond->initialized == 0) {
+		mutexLock(pthread_common.mutex_cond_init_lock);
+		if (cond->initialized == 0) {
+			err = condCreate(&cond->condh);
+			if (err == EOK) {
+				cond->initialized = 1;
+			}
+		}
+		mutexUnlock(pthread_common.mutex_cond_init_lock);
+	}
+	if (err == EOK) {
+		err = condBroadcast(cond->condh);
+	}
+	return err;
 }
 
 
 int pthread_cond_wait(pthread_cond_t *restrict cond, pthread_mutex_t *restrict mutex)
 {
-	int err = condWait(*cond, *mutex, 0);
+	int err = EOK;
 
-	while (err == -EINTR) {
-		err = mutexLock(*mutex);
-		if (err == EOK) {
-			err = condWait(*cond, *mutex, 0);
+	if (cond->initialized == 0) {
+		mutexLock(pthread_common.mutex_cond_init_lock);
+		if (cond->initialized == 0) {
+			err = condCreate(&cond->condh);
+			if (err == EOK) {
+				cond->initialized = 1;
+			}
+		}
+		mutexUnlock(pthread_common.mutex_cond_init_lock);
+	}
```

Comment:
How about refactoring the common part of lazy resource allocation? Right now there is code duplication between this and other functions that can cause lazy allocation.